include/utl/logging.h

```diff
@@ -57,46 +45,83 @@ inline std::string now() {
   return ss.str();
 }
 
-/// Produce a new log line at the given `level`, with the given prefix `ctx` and
-/// message
-template <typename... Args>
-void log(log_level const level, char const* ctx,
-         fmt::format_string<Args...> fmt_str, Args&&... args) {
-  if (level >= ::utl::s_verbosity) {
-    fmt::print(std::clog, "{time} [{level}] [{ctx}] {msg}\n",
-               fmt::arg("time", now()), fmt::arg("level", to_str(level)),
-               fmt::arg("ctx", ctx),
-               fmt::arg("msg", fmt::format(fmt::runtime(fmt_str),
-                                           std::forward<Args>(args)...)));
+/// Produce a new log line at the given `level`, with the given message
+template <log_level LogLevel, typename... Args>
+struct log {
+  log(const char* ctx, fmt::format_string<Args...> fmt_str,
+      const Args&&... args,
+      std::source_location const& loc = std::source_location::current())
+      : loc_{loc},
+        ctx_{ctx},
+        msg_{fmt::format(fmt_str, std::forward<Args>(args)...)} {}
+
+  ~log() {
+    if (LogLevel >= utl::s_verbosity) {
+#if defined(_WIN32)
+      const char* baseFileName = strrchr(loc_.file_name(), '\\')
+                                     ? strrchr(loc_.file_name(), '\\') + 1
+                                     : loc_.file_name();
+#else
+      // On MacOS, due to a bug with Clang 15, the wrong filename
+      // is retrieved (logging.h instead of the calling file):
+      // https://github.com/llvm/llvm-project/issues/56379
+      const char* baseFileName = strrchr(loc_.file_name(), '/')
+                                     ? strrchr(loc_.file_name(), '/') + 1
+                                     : loc_.file_name();
+#endif
```

Comment:
We use `snake_case` for local variables.

We try to make everything `auto`. This makes it obvious when you do implicit casts (because you would have to write `static_cast` on the right side of the assignment).
https://herbsutter.com/2013/08/12/gotw-94-solution-aaa-style-almost-always-auto/

Reply:
Oops. `auto const` to be precise.

Reply:
Alright, done! 🙂